### DISCUSSION TITLE
Convert Vec<u8> in Rust to Uint8Array in Javascript

### DISCUSSION THREAD

**Discussion by yang282441848** [tauri-apps/tauri#6286] [Ideas]
Api in Tauri covert `Vec<u8>` to `number[]` in javascript , and cover `number[]` to `Uint8Array`. There are two traverses in this step.This is's efficient way to transfer a large size of binary data from Rust to Javascript.
```
// tooling/api/src/fs.ts
async function readBinaryFile(
  filePath: string,
  options: FsOptions = {}
): Promise<Uint8Array> {
  const arr = await invokeTauriCommand<number[]>({
    __tauriModule: 'Fs',
    message: {
      cmd: 'readFile',
      path: filePath,
      options
    }
  })

  return Uint8Array.from(arr)
}
```

Is there any way to cover `Vec<u8>` to `Uint8Array` without memory copy?

**Comment by yang282441848** [comment]
I want to make a plugin to decode rtsp and render it to webview.

1.  use ffmpeg to get yuv data from rtsp. [`Rust`] 
2.  use tauri event to pass yuv data to webview. [`Rust`] -> [`webview`]
3.  use webgl to render videos.[`webview`]

There are chunk datas from `Rust` to `webview`.

**Comment by FabianLars** [comment]
> Is there any way to cover Vec<u8> to Uint8Array without memory copy?

Nope. and it gets worse than that, the binary data will be converted to a json string first because that's what the current IPC requires. (btw shouldn't converting number[] to Uint8Array be relatively cheap?).

You may have more luck with [custom protocols](https://docs.rs/tauri/latest/tauri/struct.Builder.html#method.register_uri_scheme_protocol) - examples:
https://github.com/tauri-apps/tauri/tree/dev/examples/streaming
https://github.com/JonasKruckenberg/pisano/
It should handle binary payloads much better than the IPC

**Reply by yang282441848** [reply]
Thx so much.

Why not use shared memory for `IPC`? It turns to be 2fps from 25fps after using `IPC` by `emit` api....

I will try to use `streaming`

**Reply by vdvman1** [reply]
Not all the webviews that Tauri use support shared memory

**Reply by FabianLars** [reply]
> Why not use shared memory for IPC?

Because this is not possible yet. There is no WebView API we could use for that yet, but soon ™️ as all webviews apparently have something experimental/in-preview
